### ISSUE TITLE
bn.powermod is painfully slow

### ISSUE THREAD

**Issue by ItalyPaleAle** [bitwiseshiftleft/sjcl#172]
We are implementing a full SRP system for browser-based authentication.

On the server side, we are using "mozilla/node-srp", which uses "justmoon/node-bignum" for dealing with big integers. Speed is excellent, but that is not surprising given that node-bignum uses native code.

On the client side, however, we wanted to use SJCL, as we are already using this library for other tasks in the same project.
Turns out that SJCL has already some methods implemented for SRP, and we were writing the missing ones.

In particular, we had problems with calculating `A = g^a`. This is the code:

``` js
var group = sjcl.keyexchange.srp.knownGroup(2048)
// Get random number a
var secret1 = sjcl.random.randomWords(8, 10) // 32 bytes key
// Compute srpA = g^a
var x = sjcl.bn.fromBits(secret1)
var srpA = group.g.powermod(x, group.N)
```

The call to `powermod` above takes ~7s on Chrome 34, on OSX Mavericks on a 2010 MacBook Air, 2GHz Core i7. In certain cases, the web page inside Chrome crashes, forcing me to reload it.
On Safari, same machine, the browser hangs completely.

Changing the group from 2048 to 1024 improves the situation very slightly (~5s), but not enough.

Is there anything that could be done to improve performances of the big integer part of SJCL?

**Comment by danielstockton** [comment]
@EgoAleSum What library did you use in the end?

**Comment by EgoAleSum** [comment]
JSBN for this part only

**Comment by Bren2010** [comment]
@EgoAleSum Does pull request #208 give you times on par with what you're looking for?

**Comment by heavensrevenge** [comment]
@Bren2010 The new code is ~30x faster but it still doesn't seem to be on par with the others (~6.5x slower)
http://jsperf.com/jsbn-powermod is where I tested your code to compare and http://jsperf.com/jsbnvsleemonvssjclbiginteger/13 is my origional.

**Comment by EgoAleSum** [comment]
Thank you @heavensrevenge for the JSPerf!
I've added http://jsperf.com/jsbn-powermod/2 with odd generators (see #208), and there `powermod()` (non-Montgomery) achieves basically the same speed as the montgomery one.

@Bren2010 the patch is terrific and the speed improvement is incredible. It's still far behind but seems a bit more usable now.

**Comment by Bren2010** [comment]
@EgoAleSum `powermod()` is seeing that the modulus is odd and calling `montpowermod()` instead of continuing.  That's the reason the benchmarks are the same.  The parity of the generator shouldn't matter...

@heavensrevenge @EgoAleSum I did some cleaning and implemented sliding window exponentiation.  It's still not as fast as JSBN, but it's better than it was before.  (I think it's as fast as I can get it at the moment.)

Thank you both!!

**Comment by heavensrevenge** [comment]
not bad @Bren2010, your latest changes double the speed of the original code.
If need be you could always take example or tricks from JSBN but very good work thus far.

**Comment by EgoAleSum** [comment]
Confirm that. From 0.58 ops/s on my laptop to ~25 ops/second. Still lags behind JSBN (135), but it's already 2 times or so Laemon Baird. It's certainly usable now!

Great work @Bren2010 , and it's improved even more!

**Comment by danielstockton** [comment]
:+1 nice work

**Comment by heavensrevenge** [comment]
@Bren2010 maybe you can take some ideas from https://github.com/vibornoff/asmcrypto.js/blob/release/src/bignum/modulus.js it's blazing fast when it comes to computation http://jsperf.com/jsbn-vs-bigint-js-modular-exponentiation-montgomery/4 , although its bignum creation slows it down http://jsperf.com/jsbn-powermod/3

**Comment by federicobond** [comment]
This is an old thread, but I wanted to chime in because I recently had to implement a system similar to the one described by @Bren2010 and @EgoAleSum, where the same password is used for both authentication and client-side encryption via `sjcl.encrypt`. It is certainly not optimal, but it provides a lot of convenience for end users.

While they are both right that hashing only client-side would effectively make the system equivalent to one that stores passwords in the clear, the solution they propose (hashing the password and sending that to the server) ignores the reason why we use key derivation functions like PBKDF2 instead of simple hashes. On modern hardware, it is not uncommon to be able to perform millions or even billions of hashes per second, so using the SHA-256 of the password for authentication is not much different in terms of security that using the raw password. The service provider could easily break plenty of passwords and access the encrypted payloads. The point of key derivation functions is that they are slow, and they must be because the password space is usually quite reduced, compared to hashing whole documents for example.

The proper solution is to also perform key derivation client-side on the authentication password (obviously with a different salt than the one used in the `sjcl.encrypt` payload). An alternative to obtaining the user salt by the server or using a fixed one is deterministically generating it using something else we know about the user. The only other piece of data we have available at every authentication endpoint is the username, so we can hash it to obtain the user salt. Salts do not need to be private, but they need to be different for every user (so that attackers cannot crack all passwords at the same time), so there is nothing wrong with generating them deterministically.

**Comment by EgoAleSum** [comment]
@federicobond this issue was originally opened because we were trying to implement a system based on SRP for authentication, which completely removes the need to send the password to the server: http://srp.stanford.edu/ndss.html

**Comment by federicobond** [comment]
Yes, I agree that SRP is probably a better solution. I left the comment to warn people of the pitfall I mentioned when implementing systems which share authentication and encryption passwords.

**Comment by Nilos** [comment]
From my point of view the most desirable system is one that combines SRP and a "key" derivation function because even though an SRP system does not store password equivalent data, one can still perform a brute force attack with reasonable speed on the data, whereas every "key" derivation function has a hardening factor, and one guess on a derivation function takes a long (parameterized) time compared to one guess against an SRP based system.

Combining SRP and derivation should be easy just do (pseudocode!):
password2 = derive(password, salt)
srp(password2)

**Comment by Nilos** [comment]
Still a faste bn.powermod would be desirable!

**Comment by EgoAleSum** [comment]
@Nilos not just desirable, but if you consider the results from my test above (crashing a browser) is necessary. Otherwise SRP simply can't be done

**Comment by Nilos** [comment]
@EgoAleSum I am always happy to merge a PR :)
